### DISCUSSION TITLE
Rules Based Matcher  with OP "!" and "?"

### DISCUSSION THREAD

**Discussion by galtay-tempus** [explosion/spaCy#9692] [Help: Other Questions] [answered]
## How to reproduce the behaviour

I think this is a bug, but would be interested in hearing feedback. 

```
nlp = spacy.load("en_core_web_sm")

matcher = Matcher(nlp.vocab)
patterns = [[{"TEXT": "token1", "OP": "!"}, {"TEXT": "token2", "OP": "?"}, {"TEXT": "token3"}]]
matcher.add("TEST", patterns)

doc = nlp("token1 token2 token3")
matches = matcher(doc, with_alignments=True)
print(matches)
[(1046765068364475028, 1, 3, [0, 2])]

doc = nlp("token1 token3")
matches = matcher(doc, with_alignments=True)
print(matches)
[]
```

My expectations are that the "match 0 times" token with OP "!" would cause the first example to produce no matches. 

I also think maybe this is a bug b/c indexing into the doc with start = match[1] = 1 and end=match[2] = 3 raises an IndexError. 

EDIT: 
sorry, my confusion, `doc[1:3]` works fine in the first example. 

Thanks for the great library!  

## Your Environment

- **spaCy version:** 3.2.0
- **Platform:** Linux-4.14.248-189.473.amzn2.x86_64-x86_64-with-glibc2.17
- **Python version:** 3.8.12
- **Pipelines:** en_core_web_sm (3.2.0)

**Comment by adrianeboyd** [comment]
Yes, it's the same as array slice notation, so `doc[1:3]` is correct, where the end index is exclusive.

**Comment by polm** [comment, marked as answer]
Also, to clarify:

> My expectations are that the "match 0 times" token with OP "!" would cause the first example to produce no matches.

This is not correct, the output is as expected. You can remove `token1` from the text and it will still work, because `token2` is not `token1` (matches ! rule), `token2` is present zero times after that (matches ? rule), and `token3` is next (matches final rule).

**Reply by galtay-tempus** [reply]
thanks! yes, this makes it clear, i was not interpreting the pattern i wrote correctly.
